---
title: "Lab2 732A51 Bioinformatics Group 9"
author: "Duc Duong, Martin Smelik, Raymond Sseguya"
date: "2018 M11 27"
output: pdf_document
---

##Question 1: DNA sequence acquisition and simulation
###1.1 Simulate an artificial DNA sequence dataset

```{r setup, echo=FALSE, warning=FALSE, message=FALSE,include=FALSE}
library(ape)
library(seqinr)
library(phangorn)
library(markovchain)
library(msa)
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, include=TRUE)
```

Firstly, the code block below will get the lizards DNA from GenBank, and save it as a fasta file. (code provided by the teacher)

```{r}
## Gene bank accession numbers taken from http://www.jcsantosresearch.org/Class_2014_Spring_Comparative/pdf/week_2/Jan_13_15_2015_GenBank_part_2.pdf
lizards_accession_numbers <- c("JF806202", "HM161150", "FJ356743", "JF806205", 
                               "JQ073190", "GU457971", "FJ356741", "JF806207",
                               "JF806210", "AY662592", "AY662591", "FJ356748",       
                               "JN112660", "AY662594", "JN112661", "HQ876437", 
                               "HQ876434", "AY662590", "FJ356740", "JF806214", 
                               "JQ073188", "FJ356749", "JQ073189", "JF806216", 
                               "AY662598", "JN112653", "JF806204", "FJ356747", 
                               "FJ356744", "HQ876440", "JN112651", "JF806215",
                               "JF806209") 
lizards_sequences<-ape::read.GenBank(lizards_accession_numbers)
ape::write.dna(lizards_sequences, file ="lizard_seqs.fasta", format = "fasta", append =FALSE, nbcol = 6, colsep = "", colw = 10)
print(lizards_sequences)

```

We created a function called simulate_gene, which take the lizards sequences as the input and simulate an AI gene base on the original sequences. When calling the function, AI simulated gene is created automatically, and saved in a file called `AI_gene.fasta`. A message is returned to announce that the file is saved successfully.

```{r}
#1.1
clean <- function(template_gene){
  nucleotide <- c("a", "t", "g", "c")
  for (i in 1:length(template_gene)) {
    #Remove the " " that created when reading a file
    template_gene[[i]] <- template_gene[[i]][template_gene[[i]]!= " "]
    #Remove the character that not nucleotide (eg: name of species...)
    template_gene[[i]] <- template_gene[[i]][match(template_gene[[i]], nucleotide)]
  }
  return(template_gene)
}

# Read and clean 
lizards_sequences <-  read.fasta("lizard_seqs.fasta")
lizards_sequences <- clean(lizards_sequences)

#Simulate AI gen function
simulate_gene <- function(template_gene)
{
  ai_gene <- list()
  gene_num <- length(template_gene)
  nucleotide <- c("a", "t", "g", "c")
  
  #Scan all gene and get some information
  for (i in 1:gene_num) {
    
    template_sequence <- template_gene[[i]]
    #get leng and base compotision of gene
    this_leng <- length(template_sequence)
    this_compotision = seqinr::count(template_sequence,1)/this_leng
    
    #generate a new sequence base on sample function
    this_sequence <- sample(nucleotide, size=this_leng ,prob = this_compotision, replace = TRUE)
    #print(this_sequence)
    
    #add to list
    ai_gene[i] <- list(this_sequence)
  }
  
  #write to a file
  ape::write.dna(ai_gene, file ="AI_gene.fasta", format = "fasta", colsep =" ")
  return("Created an AI gene and saved in file: AI_gene.fasta")
}


simulate_gene(lizards_sequences)



```

Each sequence of the AI simulated gene has the same base composition with the original gene. For example, here is the base composition of the first sequence:

```{r}

ai_gene_1.1 <- read.fasta("AI_gene.fasta")
ai_gene_1.1 <- clean(ai_gene_1.1)

print("Base composition of the first sequence of the original gene: ")
print(count(lizards_sequences[[1]],1)/length(lizards_sequences[[1]]))

print("Base composition of the first sequence of the AI gene: ")
print(count(ai_gene_1.1[[1]],1)/length(ai_gene_1.1[[1]]))

```


##1.2 Artificial DNA sequence dataset using phangorn::simSeq() function.
Here is the phylogenetic tree with 33 tips:
```{r}
#1.2
tree <- rtree(length(lizards_sequences))
plot(tree)
```

Then, we create a transition Q matrix base of random number from 0.22 to 0.28. Which quite similar with the true number of the real data.

```{r}

#the rates matrix
rates <- matrix(0, ncol = 4, nrow = 4)
rownames(rates) <- c("a", "c", "g", "t")
colnames(rates) <- c("a", "c", "g", "t")

#fill value to rates matrix
for (i in 1:4) {
  rate = runif(3, 0.22, 0.28)
  for (j in 1:4) {
    if (j==4) 
    {
      rates[i,j]= 1- sum(rate)
    }else rates[i,j] = rate[j]
  }
}

#print the rates
rates

```

Finally, we create the second AI simulated gene base on the phangorn::simSeq() function. We choose the length of all sequences equals 1000, Which more or less like the original sequence.   
The second AI simulated gene is saved as a fasta file with the name AI_gene2.fasta

```{r}
#create the ai_gene 2
ai_gene_1.2 <- phangorn::simSeq(tree, l = 1000, Q=rates , type = "DNA")

#rename 
for (i in 1:length(ai_gene_1.2)){
  ai_gene_1.2[[i]][ai_gene_1.2[[i]] == 1] = "a"
  ai_gene_1.2[[i]][ai_gene_1.2[[i]] == 2] = "c"
  ai_gene_1.2[[i]][ai_gene_1.2[[i]] == 3] = "g"
  ai_gene_1.2[[i]][ai_gene_1.2[[i]] == 4] = "t"
}

ape::write.dna(ai_gene_1.2, file ="AI_gene2.fasta", format = "fasta", colsep ="")
```


## Question2: Sequence analysis
### 2.1: Report some basic statistics
Here is some basic statistics as the requirements:
```{r}
#2.1
ai_gene_1.2 <- read.fasta("AI_gene2.fasta")
ai_gene_1.2 <- clean(ai_gene_1.2)

for (i in 1:length(lizards_sequences)) {
  cat(paste("For the sequences number: ", i , "\n"))
  print("The composition of lizards dataset:")
  print(round(count(lizards_sequences[[i]],2)/length(lizards_sequences[[i]]), 4))
  
  print("The composition of AI_gene1 dataset:")
  print(round(count(ai_gene_1.1[[i]],2)/length(ai_gene_1.1[[i]]), 4))
  
  print("The composition of Ai_gene2 dataset:")
  print(round(count(ai_gene_1.2[[i]],2)/length(ai_gene_1.2[[i]]), 4))
  
  cat("\n")
}


```

###2.2: Markov chain
We decided to use the `markovchain` library to fit the markovchanin model. Here is the resutl:

```{r}
library(markovchain)
markovchainFit(lizards_sequences)
markovchainFit(ai_gene_1.1)
markovchainFit(ai_gene_1.2)
```

Markov chain order: I'll do it later. There is an error in here

```{r, eval=FALSE}
#2.2
#The fitHigherOrder function work only with a list, not list of list
#So, my idea is take some random sequences, see the order and then make the conclusion 
fitHigherOrder(lizards_sequences[[1]])

```

###2.3: Align the sequences

We use the `msa` library to align the sequences, then calculate the distantce and draw some heatmaps as below:

```{r}
#2.3
library(msa)

real_align <- msaClustalW("lizard_seqs.fasta",type="dna") 
real_alignseq<- msaConvert(real_align, type="seqinr::alignment") 
dist_real <- as.matrix(dist.alignment(real_alignseq, "identity"))
heatmap(dist_real)


ai1.1_align <- msaClustalW("AI_gene.fasta",type="dna") 
ai1.1_alignseq <- msaConvert(ai1.1_align, type="seqinr::alignment") 
dist_a1.1 <- as.matrix(dist.alignment(ai1.1_alignseq, "identity"))
heatmap(dist_a1.1)

ai1.2_align <- msaClustalW("AI_gene2.fasta",type="dna") 
ai1.2_alignseq<- msaConvert(ai1.2_align, type="seqinr::alignment") 
dist_a1.2 <- as.matrix(dist.alignment(ai1.2_alignseq, "identity"))
heatmap(dist_a1.2)


```
 
As we can see that the values in heatmap of AI simulated sequences are quite low. It means that the AI simulated gene has low (or no) connections with each other (because it randomly created) while the original gene is highly connected.

##Question 3: Phylogeny reconstruction

```{r include=TRUE, eval=TRUE, warning=FALSE, message=FALSE, echo=FALSE}

#3.1
treeUPGMA <- upgma(dist_real)
treeNJ <- upgma(dist_real)
layout(matrix(c(1,2), 2, 1), height=c(1,2))
par(mar = c(0,0,2,0)+ 0.1)
plot(treeUPGMA, main="UPGMA")
plot(treeNJ, "unrooted", main="NJ")

#haven't done the last question


```

